crates/sui-core/src/consensus_throughput_calculator.rs

```diff
@@ -0,0 +1,517 @@
+            .1
+    }
+    /// Resolves the traffic profile that corresponds to the provided throughput.
+    pub fn resolve(&self, current_throughput: u64) -> TrafficProfile {
+        let mut iter = self.profiles.iter();
+        while let Some((threshold, profile)) = iter.next_back() {
+            if current_throughput >= *threshold {
+                return *profile;
+            }
+        }
+
+        warn!("Could not resolve traffic profile for throughput {} - we shouldn't end up here. Fallback to lowest profile as default.", current_throughput);
+
+        // If not found, then we should return the lowest possible profile as default.
+        self.lowest_profile()
+    }
+}
+
+impl Default for TrafficProfileRanges {
+    fn default() -> Self {
+        let profiles = vec![
+            TrafficProfile {
+                level: Level::Low,
+                throughput: 0,
+            },
+            TrafficProfile {
+                level: Level::High,
+                throughput: 2_000,
+            },
+        ];
+        TrafficProfileRanges::new(&profiles)
+    }
+}
+
+pub type TimestampSecs = u64;
+
+#[derive(Debug)]
+pub struct TrafficProfileEntry {
+    /// The traffic profile
+    profile: TrafficProfile,
+    /// The time when this traffic profile was created
+    timestamp: TimestampSecs,
+    /// The calculated throughput when this profile created
+    #[allow(unused)]
+    throughput: u64,
+}
+
+#[derive(Default)]
+struct ConsensusThroughputCalculatorInner {
+    observations: VecDeque<(TimestampSecs, u64)>,
+    total_transactions: u64,
+}
+
+pub struct ConsensusThroughputCalculator {
+    /// The number of transaction traffic observations that should be stored within the observations
+    /// vector in the ConsensusThroughputCalculatorInner. Those observations will be used to calculate
+    /// the current transactions throughput. We want to select a number that give us enough observations
+    /// so we better calculate the throughput and protected against spikes. A large enough value though
+    /// will make us less reactive to traffic changes.
+    observations_window: u64,
+    /// The time that should be passed between two consecutive traffic profile updates. For example, if
+    /// we switch at point T to profile "Low", there will need to be passed at least `traffic_profile_update_window`
+    /// seconds until the traffic profile gets updated to a different value.
+    traffic_profile_update_window: TimestampSecs,
+    /// When current calculated throughput (A) is lower than previous, and the assessed profile is now a lower than previous,
+    /// we'll change to the lower profile only when (A) <= (previous_profile.throughput) * (100 - traffic_profile_cool_down_throughput_threshold) / 100.
+    /// Otherwise we'll stick to the previous profile. We want to do that to avoid any jittery behaviour that alternates between two profiles.
+    traffic_profile_cool_down_throughput_threshold: u64,
```

Comment:
Why do we need both the update min interval and cool down threshold?

Reply:
Those are basically guaranteeing two different things:
* min internal ensures that we are not changing profiles _too_ frequently. We want to respect some minimum time before we are ready to evaluate for a new profile.
* the min delay on its own doesn't though protect us from the case where the current throughput is very borderline between two traffic profiles and we can find our selfs after every min internal period alternating between those profiles - which is also undesired. For that we need the cool down period to avoid constantly fluctuating. Please check the PR description for more info - I've added more details there and experiment results:

 
<img width="846" alt="Screenshot 2023-09-28 at 16 21 13" src="https://github.com/MystenLabs/sui/assets/17335598/d3156d1d-3115-47c1-94e5-d8d9c5996f3a">